File: Algorithms/Java/Sort/SelectionSort.java
/**
 * 选择排序
 *
 * @description: 每一次选择还没处理元素里最小的元素
 * @author: WenRuo
 * @date: 2021/10/20 20:01
 */

public class SelectionSort {

    private SelectionSort() {
    }

    /**
     * 升序 Ascending order
     *
     * @param arr
     */
    public static <E extends Comparable<E>> void sort(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int minIndex = i;
            for (int j = i; j < arr.length; j++) {
                if (arr[j].compareTo(arr[minIndex]) < 0)
                    minIndex = j;
            }
            swap(arr, i, minIndex);
        }
    }

    /**
     * 降序 Descending order
     *
     * @param arr
     */
    public static <E extends Comparable<E>> void sort2(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int minIndex = i;
            for (int j = i; j < arr.length; j++) {
                if (arr[j].compareTo(arr[minIndex]) > 0)
                    minIndex = j;
            }
            swap(arr, i, minIndex);
        }
    }

    private static <E> void swap(E[] arr, int i, int j) {
        E tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
}

File: Algorithms/C#/Sort/SelectionSort.cs
//
// @Description:
// @Author: WenRuo
//


using System;

namespace Sort
{
    public class SelectionSort
    {
        private SelectionSort()
        {
        }

        /// <summary>
        /// 升序
        /// </summary>
        /// <param name="arr"></param>
        /// <typeparam name="T"></typeparam>
        public static void Sort<T>(T[] arr) where T : IComparable
        {
            for (int i = 0; i < arr.Length; i++)
            {
                int minIndex = i;
                for (int j = i; j < arr.Length; j++)
                {
                    if (arr[j].CompareTo(arr[minIndex]) < 0)
                        minIndex = j;
                }

                Swap(arr, i, minIndex);
            }
        }

        /// <summary>
        /// 降序
        /// </summary>
        /// <param name="arr"></param>
        /// <typeparam name="T"></typeparam>
        public static void Sort2<T>(T[] arr) where T : IComparable
        {
            for (int i = 0; i < arr.Length; i++)
            {
                int minIndex = i;
                for (int j = i; j < arr.Length; j++)
                {
                    if (arr[j].CompareTo(arr[minIndex]) > 0)
                        minIndex = j;
                }

                Swap(arr, i, minIndex);
            }
        }

        private static void Swap<T>(T[] arr, int i, int j)
        {
            T tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
    }
}